rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleParser.java

```diff
@@ -52,49 +45,39 @@ public class GradleParser implements Parser<G.CompilationUnit> {
+                    "gradle-testing-base", "gradle-testing-jvm", "gradle-enterprise-gradle-plugin");
         } catch (IllegalArgumentException e) {
             // when gradle API has been fatjared into the rewrite-gradle distribution
-            this.groovyParser = groovyParser.build();
         }
+        this.buildParser = GroovyParser.builder(base)
+                .compilerCustomizers(
+                        config -> config.setScriptBaseClass("RewriteGradleProject")
+                )
+                .build();
+        this.settingsParser = GroovyParser.builder(base)
+                .compilerCustomizers(
+                        config -> config.setScriptBaseClass("RewriteSettings")
+                )
+                .build();
     }
 
     @Override
     public List<G.CompilationUnit> parseInputs(Iterable<Input> sources, @Nullable Path relativeTo, ExecutionContext ctx) {
-        Iterable<Input> gradleWrapped = StreamSupport.stream(sources.spliterator(), false)
-                .map(source ->
-                        new Parser.Input(
-                                source.getPath(), source.getFileAttributes(),
-                                () -> new SequenceInputStream(
-                                        Collections.enumeration(asList(
-                                                new ByteArrayInputStream(source.getPath().endsWith(Paths.get("settings.gradle")) ?
-                                                        SETTINGS_PREAMBLE : PREAMBLE),
-                                                source.getSource(ctx),
-                                                new ByteArrayInputStream(new byte[]{'}', '}'})
-                                        ))
-                                ),
-                                source.isSynthetic()
-                        )
-                )
-                .collect(Collectors.toList());
-
-        return groovyParser.parseInputs(gradleWrapped, relativeTo, ctx).stream()
-                .map(cu -> {
-                    List<Statement> projectBody = cu.getClasses()
-                            .get(cu.getClasses().size() - 1)
-                            .getBody().getStatements();
-                    J.MethodDeclaration script = (J.MethodDeclaration) projectBody.get(projectBody.size() - 1);
-                    assert script.getBody() != null;
-                    return cu.withStatements(script.getBody().getStatements());
-                })
-                .collect(Collectors.toList());
+        return StreamSupport.stream(sources.spliterator(), false)
+              .flatMap(source -> {
+                  if (source.getPath().endsWith("settings.gradle")) {
+                      return settingsParser.parseInputs(Collections.singletonList(source), relativeTo, ctx).stream();
+                  }
+                  return buildParser.parseInputs(Collections.singletonList(source), relativeTo, ctx).stream();
```

Comment:
I'm assuming that order of source inputs must be maintained, so having to parse files as they come through based on what type each Gradle file is.